### ISSUE TITLE
Write verilog doesn't reflect interconnected power ports

### ISSUE THREAD

**Issue by smunaut** [parallaxsw/OpenSTA#27]
You can have multiple power / ground ports connected to each other in the database and this will not be reflected in the written verilog.
This results in LVS errors later in the flow because the powered netlist written by openroad doesn't math the extracted spice netlist.

I applied this path to fix it locally :

```difff
diff --git a/verilog/VerilogWriter.cc b/verilog/VerilogWriter.cc
index e3362da..51cf317 100644
--- a/verilog/VerilogWriter.cc
+++ b/verilog/VerilogWriter.cc
@@ -214,9 +214,9 @@ VerilogWriter::verilogPortDir(PortDirection *dir)
   else if (dir == PortDirection::bidirect())
     return "inout";
   else if (dir == PortDirection::power())
-    return "input";
+    return "inout";
   else if (dir == PortDirection::ground())
-    return "input";
+    return "inout";
   else if (dir == PortDirection::internal())
     return nullptr;
   else {
@@ -412,7 +412,7 @@ VerilogWriter::writeAssigns(Instance *inst)
     Net *net = network_->net(term);
     Port *port = network_->port(pin);
     if (port
-        && network_->direction(port)->isAnyOutput()
+        && (network_->direction(port)->isAnyOutput() || network_->direction(port)->isPowerGround())
         && !stringEqual(network_->name(port), network_->name(net))) {
       // Port name is different from net name.
       string port_vname = netVerilogName(network_->name(port),
```

**Comment by jjcherry56** [comment]
makes sense. fixed in
fdca0dff write_verilog use inout for power/ground

**Comment by smunaut** [comment]
There is actually an issue I didn't originally think about but if the net that those power ports are connected to is not named like any external ports, then you get the `assign` written properly but there is never any `wire` declaration for them. Seems that when writing the `wire` declarations, the power nets are never even considered ?

**Comment by smunaut** [comment]
This got broken again at some point ...

**Comment by jjcherry56** [comment]
"This" is too nebulous to be actionable.
File a new issue **with a test case** if you expect any response.

**Comment by smunaut** [comment]
I have now traced it to OpenROAD removing power pins from the PinIterator meaning ... nothing OpenSTA can do about it if the pins aren't there in the first place. I opened an issue on OpenROAD with the test case ( https://github.com/The-OpenROAD-Project/OpenROAD/issues/10414 ) and waiting to see what they say.

I have in the mean time written a patch for OpenSTA that goes through the portIterator to get those assignement made. I don't necessarily think it's the right thing to do and waiting what OpenROAD has to say on the best way to tackle the issue, but in the mean time, that fixed my LVS problem and allows to build chips for ChipFoundry.

https://github.com/TinyTapeout/librelane/blob/da79f3ac940cd4ce345c43c5c118b679bcfa006c/nix/patches/opensta/fix_power_pins.patch
